### DISCUSSION TITLE
USWDS 3.0.0 Beta 2

### DISCUSSION THREAD

**Discussion by thisisdano** [uswds/uswds#4452] [General]
# What's new in 3.0.0 Beta 2

This release builds on [Beta 1](https://github.com/uswds/uswds/releases/tag/v3.0.0-beta.1) and is up-to-date with [USWDS 2.12.2](https://github.com/uswds/uswds/releases/tag/v2.12.2).

This release is focused on component lifecycles for any component enhanced with JavaScript (like Combo Box, Date Picker, and Modal). So, if you integrate these components into a JavaScript application and have been looking for some more flexibility with component initialization, check it out.

JavaScript components now allow for references other than` document.body` to be referenced with initialization. This makes make the JavaScript-enhanced components more flexible and easy to use, especially in scenarios where components are rendered after initial page load + USWDS init, as with Angular, React, jQuery, vanilla JavaScript, etc.

For example, passing a selector to the Modal that is _not_ `document.body`:

```
  const modalRef = useRef()

  useEffect(() => {
    const modalElement = modalRef.current;
    modal.on(modalElement)
    return () => {
      modal.off(modalElement)
    }
  }, [modalRef])
```

## What's in this release
- [x] Drop support for IE11 [IN PROGRESS]
- [x] Reorganize internal directory structure [IN PROGRESS]
- [ ] ~Use Sass Module Syntax~ [Coming in a later release]
- [ ] ~Publish component packages~ [Coming in a later release]

## Updating from stable
Install on your project with `npm install --save uswds@beta`

If you're using `includePaths` in sass, you'll need to update the path from `./node_modules/uswds/dist/scss` to ` /node_modules/uswds/dist/scss/stylesheets`.

Example using USWDS-Gulp:
```js
// OLD
.pipe(
  sass({
    includePaths: [
      `sass/styles.scss`,
      `${uswds}/scss`,
      `${uswds}/scss/packages`,
    ],
  })
)

// NEW
.pipe(
  sass({
    includePaths: [
      `sass/styles.scss`,
      `${uswds}/scss/stylesheets`,
      `${uswds}/scss/stylesheets/packages`,
    ],
  })
)
```

##### From Beta 1
_None_

##### From Stable

**Updated package names**
Only affects those using USWDS packages. The required and global partials are now prefixed with `uswds-`.

For example:
```scss
@import 'packages/required';
@import 'packages/global';
```

Required and global SCSS packages have been updated to:
```scss
@import 'packages/uswds-required';
@import 'packages/uswds-global';
```

## Security 

0 vulnerabilities in regular dependencies (dependencies for USWDS projects installed with npm install uswds)

Internal only: 53 vulnerabilities (2 low, 27 moderate, 22 high, 2 critical) in devDependencies (development dependencies)

Release ZIP SHA-256 hash: 55ffc38b87c619c9b0e6cc0d6d7bbd03b3729122f1dbc60d326ae406503dd368

<hr /><em>This discussion was created from the release <a href='https://github.com/uswds/uswds/releases/tag/v3.0.0-beta.2'>USWDS 3.0.0 Beta 2</a>.</em>

**Comment by mbucknell** [comment]
I'm confused by the statement that this version now allows initialization of components on references on elements other than document.body. We have been able to do this with the previous version. See https://code.usgs.gov/wma/iow/waterdataui/-/blob/main/assets/src/scripts/d3-rendering/info-tooltip.js . We also do the same thing for date picker and the modal (the modal is in a non public project but uses Vue and initializes the modal in the mounted lifecycle. I do have to call "init" and not on for that one. Perhaps its an issue with what I am importing in the javascript. I import from node_modules/uswds/src/components/ in USWDS 2.x . The path seems to be a bit different in USWDS 3.0 but that is the only difference I noted.

**Reply by thisisdano** [reply]
@scottqueen-bixal Maybe you can help both me and @mbucknell understand what's new in this release?

**Reply by mejiaj** [reply]
Just chiming in here, I think the problems occurred when there were multiple components (like accordions) on the same page. Toggling one would trigger another.

**Reply by mbucknell** [reply]
That would happen if you used document.body as the parameter to the component's init. Which is why we used a dom element that was scoped to just the component we wanted to initialize. But we have done that (and are doing it) with USWDS 2.x

**Reply by mbucknell** [reply]
Just to clarify, that was the exact situation (the one James described) we had except it was with tooltips. We had some that were in the DOM from the beginning and some that were dynamically created. The solution was to call init scoped to the created component.
